src/engine.rs

```diff
@@ -219,188 +195,58 @@ impl FileEngineInner {
-        for slot in 0..SLOTS_COUNT {
-            let memtables = self.memtables[slot].read().unwrap();
-            for memtable in memtables.values() {
-                // Total size of entries for this region exceed limit.
-                if memtable.entries_size() > region_entries_size_limit {
-                    info!(
-                        "region {}'s total raft log size {} exceed limit \
-                         need force compaction",
-                        memtable.region_id(),
-                        memtable.entries_size(),
-                    );
-                    regions.insert(memtable.region_id());
-                    continue;
+                let mut ents = Vec::with_capacity(entries_count);
+                let mut ents_idx = Vec::with_capacity(entries_count);
+                memtable.fetch_all(&mut ents, &mut ents_idx);
+                let mut all_ents = Vec::with_capacity(entries_count);
+                for ei in ents_idx {
+                    let e = self.read_entry_from_file(&ei, Some(&mut cache)).unwrap();
+                    all_ents.push(e);
                 }
+                all_ents.extend(ents.into_iter());
+                log_batch.add_entries(region_id, all_ents);
 
-                if inactive_file_num == 0 || gc_file_num == 0 {
-                    continue;
+                let mut kvs = Vec::new();
+                memtable.fetch_all_kvs(&mut kvs);
+                for (key, value) in kvs {
+                    log_batch.put(region_id, &key, &value);
                 }
 
-                let min_file_num = match memtable.min_file_num() {
-                    Some(file_num) => file_num,
-                    None => continue,
-                };
-
-                // Has entries left behind too far, this happens when
-                // some followers left behind for a long time.
-                if min_file_num < gc_file_num {
-                    info!(
-                        "region {}'s some followers left behind too far, \
-                         need force compaction",
-                        memtable.region_id()
-                    );
-                    regions.insert(memtable.region_id());
-                }
+                log_batches.push(log_batch);
             }
         }
-        NEED_COMPACT_REGIONS_HISTOGRAM.observe(regions.len() as f64);
-
-        regions
-    }
-
-    #[allow(dead_code)]
-    fn evict_old_from_cache(&self) {
-        let inactive_file_num = self
-            .pipe_log
-            .files_before(self.cfg.cache_size_limit.0 as usize);
-
-        if inactive_file_num == 0 {
-            return;
-        }
-
-        for slot in 0..SLOTS_COUNT {
-            let mut memtables = self.memtables[slot].write().unwrap();
-            for memtable in memtables.values_mut() {
-                memtable.evict_old_from_cache(inactive_file_num);
-            }
+        for log_batch in log_batches {
+            self.write(log_batch, false).unwrap();
```

Comment:
How about using multiple writes to prevent a huge batch which will trigger a jitter.